File: examples/test-servers/simple-mcp-server.py
#!/usr/bin/env python3
"""
Simple MCP Server for testing purposes.
Implements basic MCP protocol over stdio.
"""

import json
import sys
import os
from typing import Dict, Any, List

class SimpleMCPServer:
    def __init__(self):
        self.tools = [
            {
                "name": "echo",
                "description": "Echo back the input text",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "text": {
                            "type": "string",
                            "description": "Text to echo back"
                        }
                    },
                    "required": ["text"]
                }
            },
            {
                "name": "add",
                "description": "Add two numbers",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "a": {"type": "number", "description": "First number"},
                        "b": {"type": "number", "description": "Second number"}
                    },
                    "required": ["a", "b"]
                }
            },
            {
                "name": "env",
                "description": "Get environment variable",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string",
                            "description": "Environment variable name"
                        }
                    },
                    "required": ["name"]
                }
            }
        ]

    def handle_request(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Handle incoming MCP request"""
        method = request.get("method", "")
        params = request.get("params", {})
        request_id = request.get("id", 1)

        try:
            if method == "tools/list":
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "result": {
                        "tools": self.tools
                    }
                }
            
            elif method == "tools/call":
                tool_name = params.get("name", "")
                arguments = params.get("arguments", {})
                
                if tool_name == "echo":
                    text = arguments.get("text", "")
                    return {
                        "jsonrpc": "2.0",
                        "id": request_id,
                        "result": {
                            "content": [
                                {
                                    "type": "text",
                                    "text": f"Echo: {text}"
                                }
                            ]
                        }
                    }
                
                elif tool_name == "add":
                    a = arguments.get("a", 0)
                    b = arguments.get("b", 0)
                    result = a + b
                    return {
                        "jsonrpc": "2.0",
                        "id": request_id,
                        "result": {
                            "content": [
                                {
                                    "type": "text",
                                    "text": f"Result: {a} + {b} = {result}"
                                }
                            ]
                        }
                    }
                
                elif tool_name == "env":
                    var_name = arguments.get("name", "")
                    var_value = os.environ.get(var_name, "Not found")
                    return {
                        "jsonrpc": "2.0",
                        "id": request_id,
                        "result": {
                            "content": [
                                {
                                    "type": "text",
                                    "text": f"Environment variable {var_name}: {var_value}"
                                }
                            ]
                        }
                    }
                
                else:
                    return {
                        "jsonrpc": "2.0",
                        "id": request_id,
                        "error": {
                            "code": -32601,
                            "message": f"Tool not found: {tool_name}"
                        }
                    }
            
            elif method == "initialize":
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "result": {
                        "protocolVersion": "2024-11-05",
                        "capabilities": {
                            "tools": {}
                        },
                        "serverInfo": {
                            "name": "simple-mcp-server",
                            "version": "1.0.0"
                        }
                    }
                }
            
            else:
                return {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "error": {
                        "code": -32601,
                        "message": f"Method not found: {method}"
                    }
                }
        
        except Exception as e:
            return {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {
                    "code": -32603,
                    "message": f"Internal error: {str(e)}"
                }
            }

    def run(self):
        """Main server loop"""
        # Send server info to stderr for debugging
        print(f"Simple MCP Server starting...", file=sys.stderr)
        print(f"PID: {os.getpid()}", file=sys.stderr)
        
        try:
            for line in sys.stdin:
                line = line.strip()
                if not line:
                    continue
                
                try:
                    request = json.loads(line)
                    response = self.handle_request(request)
                    print(json.dumps(response), flush=True)
                except json.JSONDecodeError as e:
                    error_response = {
                        "jsonrpc": "2.0",
                        "id": None,
                        "error": {
                            "code": -32700,
                            "message": f"Parse error: {str(e)}"
                        }
                    }
                    print(json.dumps(error_response), flush=True)
        except KeyboardInterrupt:
            print("Server shutting down...", file=sys.stderr)
        except Exception as e:
            print(f"Server error: {e}", file=sys.stderr)

if __name__ == "__main__":
    server = SimpleMCPServer()
    server.run()

File: examples/test-servers/simple-mcp-server.js
#!/usr/bin/env node
/**
 * Simple MCP Server for testing purposes (Node.js version).
 * Implements basic MCP protocol over stdio.
 */

const readline = require('readline');

class SimpleMCPServer {
    constructor() {
        this.tools = [
            {
                name: "greet",
                description: "Greet someone with a message",
                inputSchema: {
                    type: "object",
                    properties: {
                        name: {
                            type: "string",
                            description: "Name to greet"
                        }
                    },
                    required: ["name"]
                }
            },
            {
                name: "multiply",
                description: "Multiply two numbers",
                inputSchema: {
                    type: "object",
                    properties: {
                        a: { type: "number", description: "First number" },
                        b: { type: "number", description: "Second number" }
                    },
                    required: ["a", "b"]
                }
            },
            {
                name: "timestamp",
                description: "Get current timestamp",
                inputSchema: {
                    type: "object",
                    properties: {}
                }
            }
        ];
    }

    handleRequest(request) {
        const { method, params = {}, id = 1 } = request;

        try {
            switch (method) {
                case "tools/list":
                    return {
                        jsonrpc: "2.0",
                        id,
                        result: {
                            tools: this.tools
                        }
                    };

                case "tools/call":
                    const { name: toolName, arguments: args = {} } = params;
                    
                    switch (toolName) {
                        case "greet":
                            const name = args.name || "World";
                            return {
                                jsonrpc: "2.0",
                                id,
                                result: {
                                    content: [
                                        {
                                            type: "text",
                                            text: `Hello, ${name}! Greetings from Node.js MCP server.`
                                        }
                                    ]
                                }
                            };

                        case "multiply":
                            const a = args.a || 0;
                            const b = args.b || 0;
                            const result = a * b;
                            return {
                                jsonrpc: "2.0",
                                id,
                                result: {
                                    content: [
                                        {
                                            type: "text",
                                            text: `Result: ${a} × ${b} = ${result}`
                                        }
                                    ]
                                }
                            };

                        case "timestamp":
                            const now = new Date().toISOString();
                            return {
                                jsonrpc: "2.0",
                                id,
                                result: {
                                    content: [
                                        {
                                            type: "text",
                                            text: `Current timestamp: ${now}`
                                        }
                                    ]
                                }
                            };

                        default:
                            return {
                                jsonrpc: "2.0",
                                id,
                                error: {
                                    code: -32601,
                                    message: `Tool not found: ${toolName}`
                                }
                            };
                    }

                case "initialize":
                    return {
                        jsonrpc: "2.0",
                        id,
                        result: {
                            protocolVersion: "2024-11-05",
                            capabilities: {
                                tools: {}
                            },
                            serverInfo: {
                                name: "simple-mcp-server-js",
                                version: "1.0.0"
                            }
                        }
                    };

                default:
                    return {
                        jsonrpc: "2.0",
                        id,
                        error: {
                            code: -32601,
                            message: `Method not found: ${method}`
                        }
                    };
            }
        } catch (error) {
            return {
                jsonrpc: "2.0",
                id,
                error: {
                    code: -32603,
                    message: `Internal error: ${error.message}`
                }
            };
        }
    }

    run() {
        console.error("Simple MCP Server (Node.js) starting...");
        console.error(`PID: ${process.pid}`);

        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: false
        });

        rl.on('line', (line) => {
            line = line.trim();
            if (!line) return;

            try {
                const request = JSON.parse(line);
                const response = this.handleRequest(request);
                console.log(JSON.stringify(response));
            } catch (error) {
                const errorResponse = {
                    jsonrpc: "2.0",
                    id: null,
                    error: {
                        code: -32700,
                        message: `Parse error: ${error.message}`
                    }
                };
                console.log(JSON.stringify(errorResponse));
            }
        });

        rl.on('close', () => {
            console.error("Server shutting down...");
        });

        process.on('SIGINT', () => {
            console.error("Received SIGINT, shutting down...");
            rl.close();
            process.exit(0);
        });
    }
}

if (require.main === module) {
    const server = new SimpleMCPServer();
    server.run();
}

module.exports = SimpleMCPServer;